### PR TITLE
Add verification tests for recombination maps with msHOT

### PR DESCRIPTION
**Pull request by daniel-goldstein**
Pulled in msHOT and modified it to produce summary stats just like with the ms installation. Added an option to `mspms` that allows the specification of recombination "hotspots" identical to the msHOT cli. Added the following tests to the verification suite:
- simple sanity check test that msHOT and ms produce the same result on a simulation with uniform recombination rate
- simulation from the [msHOT documentation](https://uchicago.app.box.com/s/l3e5uf13tikfjm7e1il1eujitlsjdx13/file/68787014053) with recombination hotspots. This test checks out both on master and my other current [PR](https://github.com/tskit-dev/msprime/pull/862) that computes in physical space
- a hotspot with a recombination rate of 0
- multiple recombination hotspots across the sequence, with varying rates of recombination, some zero

### CODE REVIEW

File: data/Makefile [hunk 1]
```diff
@@ -21,11 +21,16 @@ ms:
 	cp msdir/ms_summary_stats ./
 	cp msdir/sample_stats ./
 
+mshot:
+	make -C msHOTdir
+	cp msHOTdir/msHOT ./
+	cp msHOTdir/msHOT_summary_stats ./
```
Comment (jeromekelleher) [inline data/Makefile]: msHot's version of summary_stats is probably identical to ms's, so there's no point in including it. We should probably just remove the code also, no point in keeping the files in our repo.

File: data/msHOTdir/clms [hunk 2]
```diff
@@ -0,0 +1,2 @@
+gcc -O2 -o ms  ms.c  rand1.c streec.c -lm
```
Comment (jeromekelleher) [inline data/msHOTdir/clms]: Don't think we need this file

File: data/msHOTdir/col1 [hunk 3]
```diff
@@ -0,0 +1 @@
+awk ' { print $'$1' }'
```
Comment (jeromekelleher) [inline data/msHOTdir/col1]: Don't think we need this file

File: data/msHOTdir/dist3.c [hunk 4]
```diff
@@ -0,0 +1,69 @@
+/* *********************** Probability of s segregating sites (no recombination) *********
```
Comment (jeromekelleher) [inline data/msHOTdir/dist3.c]: Don't think we need this file

File: data/msHOTdir/microsat.c [hunk 5]
```diff
@@ -0,0 +1,114 @@
+/*  microsat.c:   This function takes ms output and converts it is microsatellite
```
Comment (jeromekelleher) [inline data/msHOTdir/microsat.c]: Don't think we need this file

File: data/msHOTdir/migmat [hunk 6]
```diff
@@ -0,0 +1,5 @@
+-ma
```
Comment (jeromekelleher) [inline data/msHOTdir/migmat]: Don't think we need this file

File: msprime/cli.py [hunk 7]
```diff
@@ -121,15 +121,38 @@ def __init__(
...
         self._sample_size = sample_size
         self._num_loci = num_loci
         self._num_replicates = num_replicates
         self._recombination_rate = scaled_recombination_rate
         self._mutation_rate = scaled_mutation_rate
+        self._hotspots = hotspots
         # For strict ms-compability we want to have m non-recombining loci
-        recomb_map = msprime.RecombinationMap.uniform_map(
-            num_loci, self._recombination_rate, num_loci)
+        if (hotspots is None):
```
Comment (jeromekelleher) [inline msprime/cli.py]: Brackets redundant here

File: msprime/cli.py [hunk 8]
```diff
@@ -121,15 +121,38 @@ def __init__(
...
+            recomb_map = msprime.RecombinationMap.uniform_map(
+                num_loci, self._recombination_rate, num_loci)
+        else:
+            assert(len(hotspots) > 0)
```
Comment (jeromekelleher) [inline msprime/cli.py]: Brackets redundant here (assert isn't a function)

File: data/Makefile (hunk 1 above)
Comment (daniel-goldstein) [reply data/Makefile]: do you mean sample stats? `msHOT_summary_stats` is compiled with the necessary flags to produce stats for verification.

Comment (jeromekelleher) [reply data/Makefile]: yes I do, my bad.

File: data/Makefile [hunk 9]
```diff
@@ -21,11 +21,16 @@ ms:
 	cp msdir/ms_summary_stats ./
 	cp msdir/sample_stats ./
 
+mshot:
```
Comment (jeromekelleher) [inline data/Makefile]: Looks like the target should be msHot to match the filename

File: data/msHOTdir/params [hunk 10]
```diff
@@ -0,0 +1 @@
+  -t 3.0 -s 12
```
Comment (daniel-goldstein) [inline data/msHOTdir/params]: still a couple files left to clean up including this one

File: data/msHOTdir/seedms [hunk 11]
```diff
@@ -0,0 +1 @@
+40859 29729 63930
```
Comment (jeromekelleher) [inline data/msHOTdir/seedms]: Don't need this file

File: data/msHOTdir/stats.c [hunk 12]
```diff
@@ -0,0 +1,74 @@
+/*  stats.c : Calculates mean, standard deviation for a list of numbers.
```
Comment (jeromekelleher) [inline data/msHOTdir/stats.c]: Or this file

File: data/msHOTdir/tajd.c [hunk 13]
```diff
@@ -0,0 +1,120 @@
+/************************* tajima.c *************************************************************
+ This program calculates Tajima's D when number of sequences, number of segregating sites,
```
Comment (jeromekelleher) [inline data/msHOTdir/tajd.c]: Or this one

**Comment by jeromekelleher** [comment]
Excellent! One more update and I think we're good to merge. Can we 

- Add an extra  plot for the tests where we plot the density of breakpoints along the chromosome for msprime and the ms-like instance? We can do this for all ms instances. Basically, we want to see if the breakpoints are falling the right places.
- Add a few more hotspot tests where we try out variations with different rates at a number of places along the genome. We want to see peaks of breakpoints of different heights along the genome; this will be a really good check on whether we're getting the rate calculations right.

**Comment by daniel-goldstein** [comment]
Will do. Those will be really nice to have!

**Comment by daniel-goldstein** [comment]
Histograms are looking pretty good. Here's the output for the `msdoc-recomb-ex` (uniform recomb rate) and `mshot-high-recomb-variance` tests respectively:
coalescent_breakpoints

variable recombination rate coalescent_breakpoints

File: msprime/cli.py [hunk 14]
```diff
@@ -161,6 +167,35 @@ def __init__(
         self._mutation_generator = msprime.MutationGenerator(
             self._random_generator, self._mutation_rate)
 
+    def hotspots_to_recomb_map(self, hotspots, recomb_rate, num_loci):
```
Comment (jeromekelleher) [inline msprime/cli.py]: We're not using ``self`` here at all, so this is more naturally a function I think. The function should have a docstring defining the input and outout. Then we can add a few unit tests in the test_cli module
```python
class TestHotspotsToRecombMap(..)
     def test_one_hotspot(...)

```
Just a few different tests to make sure we're getting all the corner cases in for this function.

**Comment by jeromekelleher** [comment]
Excellent, great work @daniel-goldstein. It would be nice to see a bit more clearly the two distributions. Perhaps seaborn would help here? Maybe a kernel density smoothing would give us two lines we could more easily compare? (But these introduce noise too, so I'm not too sure). See also above comment on adding unit tests for the hotspot_to_recomb_map function.

**Comment by daniel-goldstein** [comment]
Here's what the same tests would look like using Seaborn's kernel density estimate. We can restrict how loosely it estimates the breakpoints equivalently to adjusting the number of bins in the histogram. Not smoothing at all is pretty inscrutable. These plots are using the default heuristics for fitting the curves:

coalescent_breakpoints

recomb map coalescent_breakpoints

File: msprime/cli.py [hunk 15]
```diff
@@ -121,15 +161,21 @@ def __init__(
...
         self._mutation_rate = scaled_mutation_rate
         # For strict ms-compability we want to have m non-recombining loci
-        recomb_map = msprime.RecombinationMap.uniform_map(
-            num_loci, self._recombination_rate, num_loci)
+        if hotspots is None:
+            self._recomb_map = msprime.RecombinationMap.uniform_map(
+                num_loci, self._recombination_rate, num_loci)
+        else:
+            self._recomb_map = hotspots_to_recomb_map(
+                    hotspots, scaled_recombination_rate, num_loci)
```
Comment (jeromekelleher) [inline msprime/cli.py]: probably better to use ``self._recombination_rate`` so the reader isn't wondering why the two cases are different.

File: msprime/cli.py [hunk 16]
```diff
@@ -628,6 +681,7 @@ def get_mspms_parser(error_handler=None):
         "--gene-conversion", "-c", type=float, nargs=2, default=(0, 1),
         metavar=("gc_recomb_ratio", "track_length"),
         help=mscompat_gene_conversion_help)
+    group.add_argument("--hotspots", "-v", type=float, nargs="+", default=None)
```
Comment (jeromekelleher) [inline msprime/cli.py]: Needs some help text. Can say something like "Recombination hotspots defined according to msHOT format. This is blah blah``

File: tests/test_cli.py [hunk 17]
```diff
@@ -296,6 +313,49 @@ class CustomExceptionForTesting(Exception):
     """
 
 
+class TestHotspotsToRecombMap(TestCli):
+
+    def verify_map(self, recomb_map, expected_positions, expected_rates):
```
Comment (jeromekelleher) [inline tests/test_cli.py]: Perfect.

**Comment by jeromekelleher** [comment]
Awesome, merging. Thanks @daniel-goldstein, this is great stuff!
